src/data_distribution/first_touch_distribution.hpp

```diff
@@ -0,0 +1,135 @@
+/**
+ * @file
+ * @brief This file implements the first-touch data distribution
+ * @copyright Eta Scale AB. Licensed under the Eta Scale Open Source License. See the LICENSE file for details.
+ */
+
+#ifndef argo_first_touch_distribution_hpp
+#define argo_first_touch_distribution_hpp argo_first_touch_distribution_hpp
+
+#include <pthread.h>
+
+#include "base_distribution.hpp"
+
+/**
+ * @note backend.hpp is not included here as it includes global_ptr.hpp,
+ *       which means that the data distribution definitions precede the
+ *       backend definitions, and that is why we need to forward here.
+ */
+namespace argo {
+	/* forward argo::backend function declarations */
+	namespace backend {
+		node_id_t node_id();
+		std::size_t& backing_offset();
+		/* forward argo::backend::atomic function declarations */
+		namespace atomic {
+			void _store_public_dir(const void* desired,
+				const std::size_t size, const std::size_t rank, const std::size_t disp);
+			void _store_local_dir(const std::size_t desired,
+				const std::size_t rank, const std::size_t disp);
+			void _load_local_dir(void* output_buffer,
+				const std::size_t rank, const std::size_t disp);
+			void _compare_exchange_dir(const void* desired, const void* expected, void* output_buffer,
+				const std::size_t size, const std::size_t rank, const std::size_t disp);
+		} // namespace atomic
+	} // namespace backend
+} // namespace argo
+
+namespace argo {
+	namespace data_distribution {
+		/**
+		 * @brief the first-touch data distribution
+		 * @details gives ownership of a page to the node that first touched it
+		 */
+		template<int instance>
+		class first_touch_distribution : public base_distribution<instance> {
+			private:
+				/**
+				 * @brief try and claim ownership of a page
+				 * @param addr address in the global address space
+				 * @return the homenode of addr
+				 */
+				static std::size_t first_touch (const std::size_t& addr);
+				/** @brief protects the owner directory */
+				pthread_mutex_t ownermutex = PTHREAD_MUTEX_INITIALIZER;
```

Comment:
I did not spot this in my initial review, but we should probably attempt to use `std::mutex` (combined with `std::lock_guard` or `std::unique_lock`) in new code instead of the pthread equivalent for portability and future-proofing (and RAII even if that may not be applicable here).